integration-tests/src/test/java/oracle/weblogic/kubernetes/ItKubernetesEvents.java

```diff
@@ -103,148 +137,392 @@
+    checkEvent(opNamespace, domainNamespace1, domainUid, DOMAIN_PROCESSING_COMPLETED, "Normal", timestamp);
+    logger.info("verify the only 1 DomainProcessing Starting/Completed event is generated");
+    assertEquals(1, getEventCount(domainNamespace1, domainUid, DOMAIN_PROCESSING_STARTING, timestamp));
+    assertEquals(1, getEventCount(domainNamespace1, domainUid, DOMAIN_PROCESSING_COMPLETED, timestamp));
+  }
 
-    withStandardRetryPolicy
-        .conditionEvaluationListener(
-            condition -> logger.info("Waiting for domain event {0} to be logged "
-                + "(elapsed time {1}ms, remaining time {2}ms)",
-                DOMAIN_PROCESSING_COMPLETED,
-                condition.getElapsedTimeInMS(),
-                condition.getRemainingTimeInMS()))
-        .until(checkDomainEvent(opNamespace, domainNamespace, domainUid,
-            DOMAIN_PROCESSING_COMPLETED,  "Normal", timestamp));
+  /**
+   * Scale the cluster beyond maximum dynamic cluster size and verify the warning event is generated.
+   */
+  @Order(6)
+  @Test
+  public void testDomainK8sEventsScalePastMax() {
+    DateTime timestamp = new DateTime(Instant.now().getEpochSecond() * 1000L);
+    try {
+      logger.info("Scaling cluster using patching");
+      String patchStr
+          = "["
+          + "{\"op\": \"replace\", \"path\": \"/spec/clusters/0/replicas\", \"value\": 3}"
+          + "]";
+      logger.info("Updating replicas in cluster {0} using patch string: {1}", cluster1Name, patchStr);
+      V1Patch patch = new V1Patch(patchStr);
+      assertTrue(patchDomainCustomResource(domainUid, domainNamespace1, patch, V1Patch.PATCH_FORMAT_JSON_PATCH),
+          "Failed to patch domain");
+
+      logger.info("verify the DomainValidationError event is generated");
+      checkEvent(opNamespace, domainNamespace1, domainUid, DOMAIN_VALIDATION_ERROR, "Warning", timestamp);
+    } finally {
+      timestamp = new DateTime(Instant.now().getEpochSecond() * 1000L);
+      logger.info("Updating domain resource to set correct replicas size");
+      String patchStr
+          = "["
+          + "{\"op\": \"replace\", \"path\": \"/spec/clusters/0/replicas\", \"value\": 2}"
+          + "]";
+      logger.info("Updating replicas in cluster {0} using patch string: {1}", cluster1Name, patchStr);
+      V1Patch patch = new V1Patch(patchStr);
+      assertTrue(patchDomainCustomResource(domainUid, domainNamespace1, patch, V1Patch.PATCH_FORMAT_JSON_PATCH),
+          "Failed to patch domain");
+
+      logger.info("verify the DomainChanged event is generated");
+      checkEvent(opNamespace, domainNamespace1, domainUid, DOMAIN_CHANGED, "Normal", timestamp);
+    }
+  }
+
+  /**
+   * Scale the cluster below minimum dynamic cluster size and verify the warning event is generated.
+   */
+  @Order(7)
+  @Test
+  public void testDomainK8sEventsScaleBelowMin() {
+    DateTime timestamp = new DateTime(Instant.now().getEpochSecond() * 1000L);
+    try {
```

Comment:
Instead of patching the domain resource, we can add the attribute allowReplicasBelowMinDynClusterSize to DomainSpec class so that it can be used in any domain creation task.

Reply:
I don't know how it might break other scaling tests since allowReplicasBelowMinDynClusterSize by default is set to true.
For this test I need to set this to false and generate the validation error.

Reply:
I want an update to DomainSpec class, so that we can optionally set this parameter to false if needed. 

Domain domain = new Domain()
            .apiVersion(DOMAIN_API_VERSION)
            .kind("Domain")
            .metadata(new V1ObjectMeta()
                    .name(domainUid)
                    .namespace(domNamespace))
            .spec(new DomainSpec()
                    .allowReplicasBelowMinDynClusterSize(false)

Ok. Since I am working on a similar usecase, I will update the DomainSpec to support this parameter.